### PR TITLE
Adding APICTL+Git Integration documentation

### PR DESCRIPTION
**Pull request by malinthaprasan**
## Purpose
Adding documents for APICTL with GIT integration feature.
Fixes: https://github.com/wso2/product-apim-tooling/issues/401

This is adding changes to the existing documentation : [CI/CD with WSO2 API Manager](https://apim.docs.wso2.com/en/next/learn/api-controller/ci-cd-with-wso2-api-management/)

The added changes are highlighted with RED rectangles.

learn-api-controller-ci-cd-with-wso2-api-management-2020-07-24

### CODE REVIEW

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 1]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
+    !!! example
+        ```bash
+        $ apictl vcs deploy -e prod
+        Everything is up-to-date
+        ```
+
+    !!! tip
+        If you have multiple APIs to promote through CI/CD, you can commit those API projects as well to same repository. The rest of the process will be the same as above.
+ 
+        If you have export (or initiallized) and commited another API **Pizzashack-1.0.0** to the same repository and then run `vcs status` and `vcs deploy` commands, you will be able to see the below outputs.
```
Comment (Chamindu36) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:259]: Can we change this as "

Comment (Chamindu36) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:259]: Can we change this as If you have export (or initialized) and committed another API **Pizzashack-1.0.0** to the same repository and then run `vcs status` and `vcs deploy` commands, you will be able to see the below outputs. Since this has some typos.

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 2]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
+        Successfully imported API
+        ```
+
+        Here, **apictl** detects that there's a new API project is available to deploy. Without re-deploying the other unchanged API **SwaggerPetstore-1.0.0**, it will deploy only the new API.
+
+    !!! important
+        For deploying an API using `vcs deploy` command: 
+        
+        -   It is mandatory to have your API projects in a GIT based version control system.
+        -   It is mandatory to have `api_params.yaml` file inside each API Project. This is created by default when you exported an API using `export-api` or initiallized an API Project using `init`. In `api_params.yaml`. The below configuration section is used to deploy the API.
```
Comment (Chamindu36) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:282]: Can we change Initiallized as initialized?

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md (hunk 1 above)
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:259]: It's the initialized word, right? Is there any other typo?

Comment (Chamindu36) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:259]: And "committed" also

Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:259]: Ah, correct.

Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md]: Fixed

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md (hunk 2 above)
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md]: Fixed

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 3]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
 
-The **apictl** tool should be installed in the automation servers to begin the process. As the tool supports a variety of platforms, including Linux/Windows and macOS, this can be done easily. The **apictl** tool supports the process of importing API Projects via the `import-api` command. This command is mainly used in the pipeline to migrate to different environments.
+The **apictl** tool should be installed in the automation servers to begin the process. As the tool supports a variety of platforms, including Linux/Windows and macOS, this can be done easily. 
 
-1.  Import the `SwaggerPetstore` API into the production environment and test the API by running the following sample command.
+#### a. Promoting API project(s) in a GIT repository using a unified approach
+
+The repository (Version Control System) needs to be cloned into the instance that is executing the CI/CD process. From 3.2.0, **apictl** has the inbuilt support to integrate with a GIT based version control system. It gives a unified command `vcs deploy` to deploy any type of project (APIs, API Products and Apps).
```
Comment (npamudika) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:214]:
```suggestion
The repository (Version Control System) needs to be cloned into the instance that is executing the CI/CD process. From 3.2.0 onwards, **apictl** has the inbuilt support to integrate with a GIT based version control system. It gives a unified command `vcs deploy` to deploy any type of project (APIs, API Products and Apps).
```

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 4]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
+
+1.  Change your working directory (`cd`) into the cloned git repository.
```
Comment (npamudika) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:216]: Should this be like below or I am not clear this sentence :)
```suggestion
1.  Change your working directory and (`cd`) into the cloned git repository.
```

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 5]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
+    If you run `vcs deploy` command again, you will see the below output indicating that the deployment is already up-to-date.
+
+    !!! example
+        ```bash
+        $ apictl vcs deploy -e prod
+        Everything is up-to-date
+        ```
+
+    !!! tip
+        If you have multiple APIs to promote through CI/CD, you can commit those API projects as well to same repository. The rest of the process will be the same as above.
```
Comment (npamudika) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:257]:
```suggestion
        If you have multiple APIs to promote through CI/CD, you can commit those API projects as well to the same repository. The rest of the process will be the same as above.
```

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 6]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
 The Automation Server can be configured to run a specific pipeline for promoting artifacts to other environments. 
 The DevOps team can develop this pipeline further to include automated tests, workflow approvals, and other tasks.  
 
-The **apictl** tool should be installed in the automation servers to begin the process. As the tool supports a variety of platforms, including Linux/Windows and macOS, this can be done easily. The **apictl** tool supports the process of importing API Projects via the `import-api` command. This command is mainly used in the pipeline to migrate to different environments.
+The **apictl** tool should be installed in the automation servers to begin the process. As the tool supports a variety of platforms, including Linux/Windows and macOS, this can be done easily. 
 
-1.  Import the `SwaggerPetstore` API into the production environment and test the API by running the following sample command.
+#### a. Promoting API project(s) in a GIT repository using a unified approach
```
[the remark this answers is not in the document]
Comment (shaniR) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:212]: Shall we reword this as Promoting APIs in a GIT repository to upper environments via CI/CD"

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 7]
```diff
@@ -294,6 +379,143 @@ Run any of the following CTL commands to get keys for the API/API Product.
...
+
+        | Field           | Description                                                                                 |
+        |-----------------|---------------------------------------------------------------------------------------------|
+        | updateApiProduct| Used to specify whether to update the API Product if it already exists during the deployment|
+        | preserveProvider| Preserve existing provider of the API Product after importing it                            |
+        | importApis      | Import the dependant API(s) along with the API Product if the dependant API(s) are not available in the target environment|
+        | updateApis      | Update the dependant API(s) in the target environment                                       |
+
+    !!! tip
+        If you have multiple API Projects to promote through CI/CD, you can commit those API Product projects as well to same repository. The rest of the process will be the same as above.
```
Comment (shaniR) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:449]: Multiple API product projects could be promoted through CI/CD by commiting them to the same repository.

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 8]
```diff
@@ -294,6 +379,143 @@ Run any of the following CTL commands to get keys for the API/API Product.
...
+
+        | Field             | Description                                                                                 |
+        |-------------------|---------------------------------------------------------------------------------------------|
+        | update            | Used to specify whether to update the application if it already exists during the deployment|
+        | preserveOwner     | Preserve existing owner of the application after importing it                               |
+        | skipSubscriptions | Specifies whether to import the subscriptions of the application                            |
+        | skipKeys          | Specifies whether to import the credentials of the application                              |
+
+    !!! tip
+        If you have multiple applications to promote through CI/CD, you can commit those applications projects as well to the same repository similar to the APIs and API Products. The rest of the process will be the same as above.
```
Comment (shaniR) [inline en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:516]: Multiple applications could be promoted through CI/CD by committing those application projects to the same repository.

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md (hunk 3 above)
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:214]: This repository is created by the user. In one of the previous steps - https://github.com/wso2/docs-apim/blame/master/en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md#L162. So it's difficult to provide an exact link. I will reword it to indicate the above. Hope that would be ok?

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md (hunk 4 above)
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:216]: Its, `cd <repo-name>` :)

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:212]: Its the unified command "vcs deploy" which can be used to deploy any project (API, App, API Product) as stated below.

But for being more clarity, +1 to the suggested topic.

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 9]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
+        ```bash
+        $ apictl vcs deploy -e prod
+        Deploying Projects (1)...
+
+        APIs (1) ...
+        1: SwaggerPetstore-1.0.0: (SwaggerPetstore-1.0.0)
+        Successfully imported API
+        ```
+
+    Now the tool will automatically detect the target environment and prepare a new artifact containing environment-related details.
```
[the remark this answers is not in the document]
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:246]: > what are the environment-related details?
Sorry, that can be removed.

The deployment will happen all the time, so accepted the 2nd option.

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 10]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
+    !!! example
+        ```bash
+        $ apictl vcs deploy -e prod
+        Everything is up-to-date
+        ```
+
+    !!! tip
+        If you have multiple APIs to promote through CI/CD, you can commit those API projects as well to same repository. The rest of the process will be the same as above.
+ 
+        If you have export (or initialized) and committed another API **Pizzashack-1.0.0** to the same repository and then run `vcs status` and `vcs deploy` commands, you will be able to see the below outputs.
```
[the remark this answers is not in the document]
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:259]: Added "Adding a new API to the GIT repository"

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md [hunk 11]
```diff
@@ -207,16 +207,101 @@ For more information on initializing an API Project using OpenAPI/Swagger Specif
...
+        Deploying Projects (1)...
+
+        APIs (1) ...
+        1: Pizzashack-1.0.0: (Pizzashack-1.0.0)
+        Successfully imported API
+        ```
+
+        Here, **apictl** detects that there's a new API project is available to deploy. Without re-deploying the other unchanged API **SwaggerPetstore-1.0.0**, it will deploy only the new API.
+
+    !!! important
```
[the remark this answers is not in the document]
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:278]: Done, with the new topic addition.

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (malinthaprasan) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:259]: Added a new section "Adding a new API to the GIT repository" with the steps.

File: en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md (hunk 4 above)
Comment (npamudika) [reply en/docs/learn/api-controller/ci-cd-with-wso2-api-management.md:216]: ahh :), shall we add that part too to sth like this?
Change your working directory and navigate into the cloned git repository (cd <repo-name>)

**Comment by Mariangela** [comment]
@shaniR , this PR is blocked due to your comments. Please approve and merge it. I checked it again and approved it, but I don't have access to merge it.
